sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java

```diff
@@ -218,125 +288,157 @@ private static void fieldDescriptorFromTableField(
-          return Long.valueOf((Integer) jsonBQValue);
-        } else if (jsonBQValue instanceof Long) {
-          return jsonBQValue;
+      case "BOOLEAN":
+      case "BOOL":
+        if (value instanceof String) {
+          return Boolean.valueOf((String) value);
+        } else if (value instanceof Boolean) {
+          return value;
         }
         break;
-      case INT32:
-        if (jsonBQValue instanceof Integer) {
-          return jsonBQValue;
+      case "BYTES":
+        if (value instanceof String) {
+          return ByteString.copyFrom(BaseEncoding.base64().decode((String) value));
+        } else if (value instanceof byte[]) {
+          return ByteString.copyFrom((byte[]) value);
+        } else if (value instanceof ByteString) {
+          return value;
         }
         break;
-      case STRING:
+      case "TIMESTAMP":
+        if (value instanceof String) {
+          try {
+            return ChronoUnit.MICROS.between(Instant.EPOCH, Instant.parse((String) value));
+          } catch (DateTimeParseException e) {
+            return ChronoUnit.MICROS.between(
+                Instant.EPOCH, Instant.ofEpochMilli(Long.parseLong((String) value)));
+          }
+        } else if (value instanceof Instant) {
+          return ChronoUnit.MICROS.between(Instant.EPOCH, (Instant) value);
+        } else if (value instanceof org.joda.time.Instant) {
+          // joda instant precision is millisecond
+          return ((org.joda.time.Instant) value).getMillis() * 1000L;
+        } else if (value instanceof Integer || value instanceof Long) {
+          return ((Number) value).longValue();
+        } else if (value instanceof Double || value instanceof Float) {
+          // assume value represents number of seconds since epoch
+          return BigDecimal.valueOf(((Number) value).doubleValue())
+              .scaleByPowerOfTen(6)
+              .setScale(0, RoundingMode.HALF_UP)
+              .longValue();
+        }
         break;
-      case DOUBLE:
-        if (jsonBQValue instanceof Double) {
-          return jsonBQValue;
-        } else if (jsonBQValue instanceof Float) {
-          return Double.valueOf((Float) jsonBQValue);
+      case "DATE":
+        if (value instanceof String) {
+          return ((Long) LocalDate.parse((String) value).toEpochDay()).intValue();
+        } else if (value instanceof LocalDate) {
+          return ((Long) ((LocalDate) value).toEpochDay()).intValue();
+        } else if (value instanceof org.joda.time.LocalDate) {
+          return Days.daysBetween(
+                  org.joda.time.Instant.EPOCH.toDateTime().toLocalDate(),
+                  (org.joda.time.LocalDate) value)
+              .getDays();
+        }
+        break;
+      case "NUMERIC":
+      case "BIGNUMERIC":
+        if (value instanceof String) {
+          return value;
+        } else if (value instanceof BigDecimal) {
+          return ((BigDecimal) value).toPlainString();
```

Comment:
We do provide encoding here:
https://github.com/googleapis/java-bigquerystorage/blob/main/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/BigDecimalByteStringEncoder.java

So that they can be pass in more efficiently. Not blocking this PR.

Reply:
done